File: server/utils/signals.go
package utils

import (
	"errors"
	"os"
	"os/signal"
	"strconv"
	"syscall"
)

type SignalHandler func(os.Signal)

const (
	SIGRTMIN = 34
	SIGRTMAX = 64
)

// Leave room for an up/down transition while the previous lifecycle action is
// still completing. os/signal uses a non-blocking send and drops a signal when
// this channel is full.
var sigChannel = make(chan os.Signal, 16)
var signalHandlers = make(map[os.Signal][]SignalHandler)

func InitSignals(signals []os.Signal) {
	signal.Notify(sigChannel, signals...)

	for _, sig := range signals {
		signalHandlers[sig] = make([]SignalHandler, 0)
	}

	go func() {
		for {
			sig := <-sigChannel

			for _, handler := range signalHandlers[sig] {
				handler(sig)
			}
		}
	}()
}

func RealTimeSignal(num int) os.Signal {
	if num < 0 || num > SIGRTMAX-SIGRTMIN {
		return nil
	}
	return syscall.Signal(SIGRTMIN + num)
}

func AddSignalHandler(signals []os.Signal, handler SignalHandler) {
	for _, sig := range signals {
		signalHandlers[sig] = append(signalHandlers[sig], handler)
	}
}

func SignalProcess(pid int, signal os.Signal) error {
	sigStr := signal.String()
	if sigStr == "" {
		return errors.New("invalid signal")
	}
	_, err := RunCommand(UseSudo, "/bin/kill", "-"+sigStr[7:], strconv.Itoa(pid))
	return err
}

File: server/utils/signals.py
import signal
import threading

signal_handlers = {}
sig_channel = []

def init_signals(signals_list):
    for sig in signals_list:
        signal_handlers[sig] = []
        signal.signal(sig, _global_handler)

def _global_handler(sig, frame):
    handlers = signal_handlers.get(sig, [])
    for handler in handlers:
        threading.Thread(target=handler, args=(sig,)).start()

def real_time_signal(num):
    # Python's signal module has signal.SIGRTMIN and signal.SIGRTMAX in 3.3+ on linux
    if hasattr(signal, 'SIGRTMIN'):
        if 0 <= num <= (signal.SIGRTMAX - signal.SIGRTMIN):
            return signal.SIGRTMIN + num
    return None

def add_signal_handler(signals_list, handler):
    for sig in signals_list:
        if sig not in signal_handlers:
            signal_handlers[sig] = []
            signal.signal(sig, _global_handler)
        signal_handlers[sig].append(handler)
